zebrad/src/commands/start/sync.rs

```diff
@@ -67,6 +74,35 @@ where
             while !self.prospective_tips.is_empty() {
                 info!("extending prospective tips");
                 self.extend_tips().await?;
+
+                // Check whether we need to wait for existing block download tasks to finish
+                while self.pending_blocks.len() > LOOKAHEAD_LIMIT {
+                    match self
+                        .pending_blocks
+                        .next()
+                        .await
+                        .expect("already checked there's at least one pending block task")
+                        .expect("block download tasks should not panic")
+                    {
+                        Ok(hash) => tracing::debug!(?hash, "verified and committed block to state"),
+                        // This is a non-transient error indicating either that
+                        // we've repeatedly missed a block we need or that we've
+                        // repeatedly missed a bad block suggested by a peer
+                        // feeding us bad hashes.
+                        //
+                        // TODO(hdevalence): handle interruptions in the chain
+                        // sync process. this should detect when we've stopped
+                        // making progress (probably using a timeout), then
+                        // continue the loop with a new invocation of
+                        // obtain_tips(), which will restart block downloads.
+                        // this requires correctly constructing a block locator
+                        // (TODO below) and ensuring that the verifier handles
+                        // multiple requests for verification of the same block
+                        // hash by handling both requests or by discarding the
+                        // earlier request in favor of the later one.
```

Comment:
This is the summary of the remaining syncer work to do, mentioned in the main PR comment. It's possible for this code to get stuck if a block we need repeatedly fails to download, but the same was true before and the fix requires fixing up some other TODOs so I'd prefer to do it in a subsequent PR.

Reply:
> ensuring that the verifier handles
> multiple requests for verification of the same block
> hash by handling both requests or by discarding the
> earlier request in favor of the later one.

At the moment, both requests are sent to the verifiers. But even if the verifiers accept the block, the state rejects duplicate block hashes, and the ChainVerifier propagates this rejection back to the caller.

In addition, the CheckpointVerifier rejects duplicates by:
  * storing up to FANOUT blocks at a single height,
  * rejecting additional blocks immediately, and
  * rejecting any other duplicates when the current range is full.

The BlockVerifier accepts duplicate blocks, but this might change when we implement contextual validation.